Improved brigadier documentation (#436)
PaperMC/docs: docs/paper/dev/api/command-api/arguments.mdx

```java title="YourPluginClass.java"
public class YourPluginClass extends JavaPlugin {
    @Override
    public void onEnable() {
        LifecycleEventManager<Plugin> manager = this.getLifecycleManager();
        manager.registerEventHandler(LifecycleEvents.COMMANDS, event -> {
            final Commands commands = event.registrar();
            commands.register(
                Commands.literal("enchantmentargumentcommand")
                .then(
                    Commands.argument("enchantmentargument", ArgumentTypes.resource(RegistryKey.ENCHANTMENT))
                        .executes(ctx -> {
                            ctx.getSource().getSender().sendPlainMessage(
                                ctx.getArgument("enchantmentargument", Enchantment.class).getKey().asString()
                            );
                            return Command.SINGLE_SUCCESS;
                        })
                ).build()
            );
        });
    }
}
```

This command has one argument of the `Enchantment` datatype. When the command is executed, the command
sender will get a message containing the key of the enchantment they selected.

> ## Advantage over string-based arguments

Reviewer: ```suggestion
## Advantages over string-based arguments
```

| MEMORY_MODULE_TYPE  | MemoryKey             | Keys for saving per-entity data                                                                                                                                                                                                                                                              |
| FROG_VARIANT        | Frog.Variant          | [Frog variants](https://minecraft.wiki/w/Frog)                                                                                                                                                                                                                                               |
| MAP_DECORATION_TYPE | MapCursor.Type        | [Types of sprites displayed on a map](https://minecraft.wiki/w/Map#Map_icons)                                                                                                                                                                                                                |
| FLUID               | Fluid                 | [Fluid types](https://minecraft.wiki/w/Fluid)     

> Brigadier itself also specifies many argument types. For more information on them, see <Javadoc name={"io.papermc.paper.command.brigadier.argument.ArgumentTypes"}>ArgumentTypes</Javadoc>

Reviewer 2: ```suggestion
Minecraft itself also specifies many argument types. For more information on them, see <Javadoc name={"io.papermc.paper.command.brigadier.argument.ArgumentTypes"}>ArgumentTypes</Javadoc>
```
I think it makes more sense this way since Brigadier itself only adds [these](https://github.com/Mojang/brigadier/tree/master/src/main/java/com/mojang/brigadier/arguments).
The arguments provided in the `ArgumentTypes` class are vanilla argument types.

Author: No, Brigadier is the command API in Minecraft, so it's just more specific to use Brigadier over Minecraft

Reviewer 2: Yeah, Brigadier is the command framework Minecraft uses, but it doesn't provide the argument types that `ArgumentTypes` gives access to.
If you use plain Brigadier, you don't have access to those arguments but only to argument types that represent basic Java types.

```java
public class IceCreamTypeArgument implements CustomArgumentType.Converted<IceCreamType, String> {

    @Override
    public @NotNull IceCreamType convert(String nativeType) throws CommandSyntaxException {
        try {
            return IceCreamType.valueOf(nativeType.toUpperCase());
        } catch (Exception e) {
            Message message = MessageComponentSerializer.message().serialize(Component.text("Invalid flavor %s!".formatted(nativeType), NamedTextColor.RED));

            throw new CommandSyntaxException(new SimpleCommandExceptionType(message), message);
        }
    }

    @Override
    public @NotNull ArgumentType<String> getNativeType() {
        return StringArgumentType.word();
    }

    @Override
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> context, SuggestionsBuilder builder) {
        for (IceCreamType flavor : IceCreamType.values()) {
            builder.suggest(flavor.name(), MessageComponentSerializer.message().serialize(Component.text("look at this cool green tooltip!", NamedTextColor.GREEN)));
        }

        return CompletableFuture.completedFuture(
                builder.build()
        );
    }
}
```

> That's a lot of code, so let's start from the top. We implemented the <Javadoc name={"io.papermc.paper.command.brigadier.argument.CustomArgumentType$Converted"}>CustomArgumentType.Converted</Javadoc>
> interface. This interface takes two type arguments: our custom enum, T, and a type that is native to 
> Minecraft, such as String, Integer, etc., called N. The native type exists so that the client can use
> the input data, as it doesn't know what our custom `IceCreamType` is.
> 
> The first method called `convert()` converts the native type (in this case `String`)
> into our own custom type called `IceCreamType`, as the name suggests. This is so we 
> can keep using our custom type internally. If we wouldn't do that, there would be no point in using a 
> custom type. If the specified input isn't found in our enum of available flavors, we
> output an error.
> 
> The second method called `getNativeType()` returns the type of string that our command argument uses.
> In our case, we're using a single word, so we return `StringArgumentType.word()`.
> 
> In the last method, `listSuggestions()`, we return `CompletableFuture<Suggestions>` so that the client
> can suggest all available options. We can even add tooltips to the suggestions to explain them in greater
> detail. 
> 
> As a last step, we need to register the Command:

Reviewer: ```suggestion
We implemented the <Javadoc name={"io.papermc.paper.command.brigadier.argument.CustomArgumentType$Converted"}>CustomArgumentType.Converted</Javadoc>
interface. This interface takes two type arguments: our custom enum, T, and a native Java type called N.

- `convert()` converts the native type (in this case `String`) into our custom type.

-  `getNativeType()` returns the type of string that our command argument uses. This uses a single word, so we return `StringArgumentType.word()`.

- `listSuggestions()` returns `CompletableFuture<Suggestions>` so that the client
can suggest all available options. We can even add tooltips to the suggestions to explain them in greater
detail. 

We then need to register the command:
```

so much of this was unnecessary waffle to me
